lib/Translation/ExportVerilog/ExportVerilog.cpp

```diff
@@ -121,6 +121,29 @@ static unsigned getPrintedIntWidth(unsigned value) {
   return stream.str().size();
 }
 
+/// Calculated the printed width of the array dims of a type.
+static size_t getPrintedTypeDimsWidth(Type type, Operation *op) {
+  int bitWidth;
+  size_t textWidth = 0;
+  if (auto array = type.dyn_cast<rtl::ArrayType>()) {
+    bitWidth = array.getSize();
+    textWidth = getPrintedTypeDimsWidth(array.getElementType(), op);
+  } else if (auto inout = type.dyn_cast<rtl::InOutType>()) {
+    return getPrintedTypeDimsWidth(getUnderlyingArrayElementType(inout), op);
+  } else {
+    bitWidth = getBitWidthOrSentinel(type);
+    if (bitWidth == -1) {
+      op->emitError("value has an unsupported verilog type ") << type;
+    } else if (bitWidth == 1) { // Width 1 is implicit.
+      return 0;
```

Comment:
Won't this do the wrong thing for a pack array of single bits?  I think you'd get:

`wire [arraysize:0] thing;`

which looks like a bit vector declaration.

Reply:
It'll do that and it's not wrong. Bit vectors == single bit arrays (which IMO is an argument for piggy-backing on std.vector).

It just so happens that SystemVerilog does a poor job distinguishing arbitrarily size integers from bit vectors. Perhaps we should adopt a naming standard -- Integers (and arrays thereof) always have a signed / unsigned specifier? Non-integers use 'bit' instead? Vice versa?

Reply:
I don't understand where you're going with this.  We want the RTL dialect to be reasonable independent of the verilog semantics, we don't want to drag its unreasonableness into our IR.

Specifically, arrays and integers are different things, so we should model them as different things in the RTL level.  I was just concerned that we are correctly lowering them to verilog, which is where we bring in its wackiness.

If lowering an array of single bits to `wire [arraysize:0] thing` syntax will always work, then I'm ok with it.

Reply:
100% agree. I'm big on strong typing and want to preserve semantic information as long as possible. Integers are semantically **not** bit vectors, SystemVerilog just doesn't differentiate between them. (And Verilog was worse.)

I was actually raising two separate issues:

- What's the difference between a "bit vector" and a single bit array? Should we distinguish between them in the RTL dialect? For that matter, what -- semantically -- is a "bit vector"? Does the name imply some of the semantics of `std.vector`?
- I would like the emitted verilog to have some clue as to what the original type was. There are a few mostly synonymous keywords in SystemVerilog to `wire`. I'm thinking that (in the future) we could use different synonyms for different originating types.

Reply:
> What's the difference..

It's about the operations that apply to them.  rtl.add applies to i32 but not to `array<32xi1>`.  Array index works with arrays but not integers etc.

> I would like the emitted verilog to have some clue as to what the original type was.

Sure, we can do something like this, perhaps even as just a comment produced by the verilog printer in the uncommon case that this comes up.

Reply:
I added type comments to module parameters. It was more than I was willing to take on tonight to add types to the existing comments elsewhere.

```
module CmpSign(
  input  [3:0] a, b,    //!firrtl.uint<4>
  input  [3:0] c, d,    //!firrtl.sint<4>
  output       out    //!firrtl.flip<uint<1>>
);
```

```
module Print(
  input       clock,    //!firrtl.clock
  input       reset,    //!firrtl.uint<1>
  input [3:0] a, b    //!firrtl.uint<4>
);
```